xla/service/gpu/transforms/cudnn_sdpa_score_mod.cc

```diff
@@ -0,0 +1,366 @@
+  fwd_hlo_to_cudnn_[fwd_comp_->parameter_instruction(0)] = attention_score;
+  // save fwd parameters tensor attributes
+  for (int i = 0; i < fwd_comp_->num_parameters(); i++) {
+    auto parameter = fwd_comp_->parameter_instruction(i);
+    fwd_parameters_.push_back(fwd_hlo_to_cudnn_[parameter]);
+  }
+  return Compile(graph, fwd_hlo_to_cudnn_, fwd_comp_);
+}
+
+Tensor ScoreModFunc::Backward(Graph graph, Tensor grad) {
+  bwd_hlo_to_cudnn_[bwd_comp_->parameter_instruction(0)] = grad;
+  // add fwd parameters tensor attribute to map
+  for (int i = 1; i < bwd_comp_->num_parameters(); i++) {
+    auto parameter = bwd_comp_->parameter_instruction(i);
+    bwd_hlo_to_cudnn_[parameter] = fwd_parameters_[i - 1];
+  }
+  return Compile(graph, bwd_hlo_to_cudnn_, bwd_comp_);
+}
+
+Tensor ScoreModFunc::Compile(
+    Graph graph,
+    absl::flat_hash_map<const xla::HloInstruction*, Tensor>& hlo_to_cudnn,
+    const xla::HloComputation* computation) {
+  VLOG(3) << "Compiling cudnn sdpa computation:\n"
+          << computation->ToString() << "\n";
+  std::vector<xla::HloInstruction*> instructions =
+      computation->MakeInstructionPostOrder();
+  for (const xla::HloInstruction* hlo : instructions) {
+    auto operand = [&hlo_to_cudnn, &hlo](int i) {
+      CHECK(hlo_to_cudnn.count(hlo->operand(i)));
+      return hlo_to_cudnn[hlo->operand(i)];
+    };
+    // check if operand i is user input to the flex graph
+    // all parameters except param 0 should return true
+    // param 0 is managed by cuDNN
+    auto is_virtual = [&hlo_to_cudnn, &hlo](int i) {
+      return hlo_to_cudnn[hlo->operand(i)]->get_is_virtual();
+    };
+
+    if (xla::HloPredicateIsOp<xla::HloOpcode::kConstant,
+                              xla::HloOpcode::kParameter>(hlo)) {
+      continue;
+    }
+    if (xla::HloPredicateIsOp<xla::HloOpcode::kBitcast,
+                              xla::HloOpcode::kBroadcast>(hlo)) {
+      hlo_to_cudnn[hlo] = operand(0);
+    } else if (xla::HloPredicateIsOp<xla::HloOpcode::kIota>(hlo)) {
+      auto iota = DynCast<xla::HloIotaInstruction>(hlo);
+      const auto attrs =
+          cudnn_frontend::graph::Pointwise_attributes()
+              .set_mode(cudnn_frontend::PointwiseMode_t::GEN_INDEX)
+              .set_compute_data_type(cudnn_frontend::DataType_t::INT32)
+              .set_axis(iota->iota_dimension())
+              .set_name(std::string(hlo->name()));
+      auto attn_score = computation->parameter_instruction(0);
+      hlo_to_cudnn[hlo] = graph->pointwise(hlo_to_cudnn[attn_score], attrs);
+      hlo_to_cudnn[hlo]->set_data_type(cudnn_frontend::DataType_t::INT32);
+    } else if (hlo->IsElementwise()) {
+      const auto compute_dtype =
+          GetComputeDataType(hlo->shape().element_type());
+      const auto mode = GetElementwiseMode(*hlo);
+      if (!mode.has_value()) {
+        LOG(FATAL) << "Unsupported elementwise operation: " << hlo->ToString()
+                   << "\n";
```

Comment:
Is this really a broken invariance? This rather sounds like an error which we should handle with an error. How about making this function return an `absl::StatusOr` and return an error here.

Reply:
Yeah. True. Makes sense.